Notify `Wallets` library about a proven fraud

lukasz-zimnoch opened the pull request:
Refs: #139 

This function allows notifying the `Wallets` library about a proven fraud committed by a wallet and cause appropriate side effects like slashing and wallet termination.

tomaszslabon commented on solidity/contracts/bridge/Wallets.sol:
```diff
@@ -299,4 +304,35 @@ library Wallets {
+    ) external {
+        Wallet storage wallet = self.registeredWallets[walletPubKeyHash];
+        require(
+            wallet.state == WalletState.Live,
```
What if the fraud is proven when the wallet is in the state `MovingFunds`?
We allow to submit a fraud challenge in this state.

lukasz-zimnoch replied:
Good catch! See https://github.com/keep-network/tbtc-v2/pull/177/commits/21bafb8624066b2cd0113fa2fd1922685c1515a4

tomaszslabon commented on solidity/contracts/bridge/Wallets.sol:
```diff
@@ -299,4 +304,36 @@ library Wallets {
+    /// @param walletPubKeyHash 20-byte public key hash of the wallet
+    /// @dev Requirements:
+    ///      - Wallet must be in Live state
+    function notifyWalletFraudProven(
+        Data storage self,
+        bytes20 walletPubKeyHash
```
In the [RFC](https://github.com/keep-network/tbtc-v2/blob/main/docs/rfc/rfc-1.adoc#fraud-proofs) there is a parameter called `fraud_slashing_amount: The amount of stake slashed from each member of a wallet for a fraud`.
We will have to pass this parameter here, so that we know how much we need to slash each operator, right?
Perhaps this parameter should be moved from the `Frauds` library to `Wallets`?

lukasz-zimnoch replied:
Yep, we can move that parameter from `Frauds` here within #133

pdyraga replied:
Please ignore this comment for now. I think I misunderstood the idea of what is the responsibility of `notifyWalletFraudProven`

pdyraga commented on solidity/contracts/bridge/Wallets.sol:
```diff
@@ -299,4 +304,36 @@ library Wallets {
+
+        emit WalletTerminated(wallet.ecdsaWalletID, walletPubKeyHash);
+
+        // TODO: Perform slashing of wallet operators and add unit tests for that.
```
Another argument for a separate function is that wallet which moved from `MovingFunds` to another state (unclear which and if we shouldn't add one more state) can be slashed if the transaction of moving funds is challenged. In this case, we will be slashing a wallet that is neither `Live` nor `MovingFunds`. At the same time, it makes only sense to terminate a wallet that is either `Live` or `MovingFunds` so the validation at line 318 makes sense.

lukasz-zimnoch replied:
See https://github.com/keep-network/tbtc-v2/pull/177#discussion_r833058231, this is why I planned to expose different public functions in `Wallets` lib to handle multiple wallet lifecycle events. That way, slashing, seizing, wallet termination, and friends can be totally hidden in the `Wallets` lib. The Bridge must only call the right function. Another benefit is reducing Bridge size since everything is kept inside `Wallets`.

pdyraga replied:
Ack, continuing the discussion in the other thread and resolving this one.

lukasz-zimnoch commented on solidity/contracts/bridge/Wallets.sol:
```diff
@@ -299,4 +304,36 @@ library Wallets {
+    /// @param walletPubKeyHash 20-byte public key hash of the wallet
+    /// @dev Requirements:
+    ///      - Wallet must be in Live state
+    function notifyWalletFraudProven(
```
Yeah, that's right but the initial idea here is similar to wallet closing (https://github.com/keep-network/tbtc-v2/pull/176), i.e. public functions for each use case and termination done within an internal function. However, I'm also open for changing it to be just `terminateWallet` and push all case-specific logic outside of `Wallets` lib if that fits better our use cases.

pdyraga replied:
I think your idea of having a per-case function in `Wallets.sol` makes sense. I would go even further and assign very specific names, e.g. `notifyWalletActionTimedOut` from #176 could be `notifyRedemptionTimeout`. 

I was confused about what is going to happen in `notifyWalletFraudProven`. I think the idea makes sense but I would change three things:
- rename `notifyWalletFraudProven` to `notifyFraud` just to simplify naming - we will have a lot of per-case functions
- make it clear in docs that `notifyFraud` executes a logic in reaction to a proven fraud and it should only be called when the fraud was confirmed
- extract the logic of termination to a separate `terminateWallet` function and call it from `notifyFraud`

Something like this:
```
    function notifyFraud(
        Data storage self,
        bytes20 walletPubKeyHash
    ) external {
        Wallet storage wallet = self.registeredWallets[walletPubKeyHash];
        require(
            wallet.state == WalletState.Live ||
                wallet.state == WalletState.MovingFunds,
            "ECDSA wallet must be in Live or MovingFunds state"
        );

        wallet.state = WalletState.Terminated;

        if (self.activeWalletPubKeyHash == walletPubKeyHash) {
            // If termination refers to the current active wallet,
            // unset the active wallet and make the wallet creation process
            // possible in order to get a new healthy active wallet.
            delete self.activeWalletPubKeyHash;
        }

        terminateWallet(...);
       
        // TODO: Perform slashing of wallet operators and add unit tests for that.
    }

function terminateWallet(...) internal {
  emit WalletTerminated(wallet.ecdsaWalletID, walletPubKeyHash);
  
  // TODO: Send termination signal to the ECDSA registry and add unit tests for that.
  //       See: https://github.com/keep-network/keep-core/issues/2864
}
```

pdyraga replied:
There is actually one more possibility to `notifyFraud`. I am looking at #176 and I see we may have functions in `Wallets` that are doing validation of a claim (e.g. `notifyWalletExhausted`) and functions that will act on an already validated claim (e.g. `notifyWalletHeartbeatFailed`). I think we should differentiate the naming e.g. `notify*` and `on*`  but we can have this discussion in #176 and in #133.

lukasz-zimnoch replied:
Yep, good idea! I'll rename it to `notifyFraud` and extract `terminateWallet` function.

lukasz-zimnoch replied:
See https://github.com/keep-network/tbtc-v2/pull/177/commits/7bc8f45e68eb7b20f880b79a56e031e4fd505efc. It looks slightly different than your proposition as the `terminateWallet` has more logic which I think is all about termination.

pdyraga replied:
Yeah, that makes sense!

lukasz-zimnoch commented:
@pdyraga I fixed the build errors. They occurred because I updated `@keep-network/ecdsa` package in https://github.com/keep-network/tbtc-v2/commit/7bc8f45e68eb7b20f880b79a56e031e4fd505efc to include `self.registry.closeWallet(...)` but in the same time, that version requires the Bridge to have `__ecdsaWalletHeartbeatFailedCallback` implemented.
